lib/max.cpp

```diff
@@ -0,0 +1,93 @@
+#include "flag_gems/operators.h"
+#include "flag_gems/utils.h"
+
+#include <iostream>
+#include "c10/cuda/CUDAStream.h"
+#include "torch/torch.h"
+#include "triton_jit/triton_jit_function.h"
+
+#include <filesystem>
+#include "ATen/WrapDimUtils.h"
+#include "ATen/native/ReduceOpsUtils.h"
+#include "c10/util/DimVector.h"
+
+
+
+namespace {
+std::tuple<at::Tensor, int64_t, int64_t> permute_reduction_axes_right(
+    const at::Tensor &tensor, at::OptionalIntArrayRef reduction_axes_opt) {
+  int64_t dim = tensor.dim();
+  c10::DimVector reduction_axes;
+
+  if (reduction_axes_opt.has_value()) {
+    reduction_axes = reduction_axes_opt.value().vec();
+  }
+
+  std::unordered_set<int64_t> reduction_set(reduction_axes.begin(), reduction_axes.end());
+
+  c10::DimVector left_axes, right_axes;
+  int64_t non_reduction_size = 1, reduction_size = 1;
+
+  for (int64_t i = 0; i < dim; ++i) {
+    if (reduction_set.count(i)) {
+      right_axes.push_back(i);
+      reduction_size *= tensor.size(i);
+    } else {
+      left_axes.push_back(i);
+      non_reduction_size *= tensor.size(i);
+    }
+  }
+
+  // Concatenate left and right axes to form the new permutation order
+  c10::DimVector permute_order = left_axes;
+  permute_order.insert(permute_order.end(), right_axes.begin(), right_axes.end());
+
+  return {tensor.permute(permute_order), non_reduction_size, reduction_size};
+}
+}  // anonymous namespace
+
+
+namespace flag_gems {
```

Comment:
Please install pre-commit in your environment to apply hooks before committing code. It enforces some code style checking and applies reformatting.